Add parser basic functions.

fernape opened the pull request:
Basic parser functions to be used as building blocks for higher level functionality.  Most of the functions return a boolean specifying if they were able to perform the action.

manadart commented on parse/parser.go:
```diff
@@ -0,0 +1,81 @@
+
+        start_pos := p.l.offset
+        token = p.l.NextToken()
+        if token.Type == IDENT {
```
I think what we want to do is define the AST insofar as our recognised input/output expressions are concerned, then apply those rules, including returning errors for malformed expressions.

Have a look at Simon's first patch where he began adding all of this, including the AST: https://github.com/canonical/sqlair/pull/1.

We won't do it quite like it is there. We'll have another expression that is essentially "string of stuff we don't care about and are passing through".

manadart commented on parse/parser.go:
```diff
@@ -0,0 +1,81 @@
+        return QueryArgument{}, nil
+}
+
+func (p *Parser) Parse() ([]QueryArgument, error) {
```
Along the same lines, see Simon's patch. The output will be a query expression, which is recursively made up of expressions.

manadart reviewed (commented):
Reasonable first shot. I think a better starting point is the AST, with expressions for input/output.

fernape committed:
Add parser basic functions.

Basic parser functions to be used as building blocks for higher level
functionality.  Most of the functions return a boolean specifying if
they were able to perform the action.

Aflynn50 commented on internal/parse/parser.go:
```diff
@@ -43,3 +45,66 @@ func (p *Parser) Parse(input string) (*ParsedExpr, error) {
+	return false
+}
+
+// isNameByte returns true if the byte passed as parameter is considered to be
```
Does this match the full namespace of SQL column names (and is it meant to?)? This doc implies it could also be something of the form \`weirdname*$^%\` https://docs.microsoft.com/en-us/sql/odbc/microsoft/column-name-limitations?view=sql-server-ver16

fernape replied:
Good point.
`isNameByte` is used (in the experiment) from `parseIdentifier` which is used to recognize things like `foo.bar` in things that come before the `AS` keyword *and* things that come after it.
Apparently SQLite [deviates a bit](https://www.sqlite.org/draft/tokenreq.html) from the standard too. Sybase IQ supports things like the pipe (`|`) in column names, postgres has its own rules, etc.
It's going to be difficult to be DB-agnostic here. Maybe we can allow for everything to be a name byte except for _blank separators_ (spaces, tabs, newlines...).

Aflynn50 replied:
That sounds reasonable, it'd probably be best to specify specifically what certainly isn't a name rather than what is. As long as the parser doesn't rely on the function too heavily

niemeyer replied:
> it'd probably be best to specify specifically what certainly isn't a name

That creates double negatives which are harder to work with ("is this not-not-part-of-a-name?").

I think the main issue here is the lack of more context: a name of what?   I suspect we don't want to use the full SQL naming convention, but rather something that will be friendlier to our environment. But again, it depends on the "name of what?" idea.

fernape replied:
Here I think of `name` as something we can use to represent:
* table name
* column name
* Go type name
* Go struct field name

So we can use it to know when to stop parsing and say, "okay, this is where the name ends". So the function is really simple in the sense that it only accepts alphanumerics and underscore. This will be a limitation mainly for table/column names since [I don't expect](https://go.dev/ref/spec#identifier) any Go code to have weird characters in the type names or struct field names. But column names can go wild in some other databases. I'm ok limiting this to some sensible defaults.

niemeyer replied:
Go names don't use underscores in general, and it may also be problematic to *stop parsing* based on that specifically, because as @Aflynn50 mentioned above, the syntax is richer than what we consider "good naming". To be clear, the constraint on good names itself is a good idea, but it's worth considering how to apply it. For the time being, I suggest dropping it and then bringing it back together with the PR that uses it, so we have a bit more context for the conversation.

fernape replied:
Removed.

niemeyer commented on internal/parse/parser_test.go:
```diff
@@ -12,3 +12,73 @@ func TestInit(t *testing.T) {
+	assert.Equal(t, false, isNameByte('à'))
+	assert.Equal(t, false, isNameByte('\n'))
+	assert.Equal(t, false, isNameByte('\r'))
+	assert.Equal(t, false, isNameByte('\t'))
```
All of these tests feel quite repetitive. It seems worth organizing them in a table format. Something like:
```
type parseHelperTest struct {
        bytef   func(byte) bool
        stringf func(string) bool
        result  bool
        data    string
}

var parseTests = []parseHelperTest{{
        {bytef: isNameByte, result: true, data: "f"}
        {bytef: isNameByte, result: true, data: "0"}
        {bytef: isNameByte, result: true, data: "5"}
        ... etc ...
}
```
Then just have one single test function that for-loops the whole thing at once.

fernape replied:
I can do like in the experiment. I couple of questions though:

1. Aren't this table-like tests more difficult to debug?
```
--- FAIL: TestRunTable (0.00s)
    parser_test.go:32: 
                Error Trace:    /parser_test.go:32
                Error:          Not equal: 
                                expected: false
                                actual  : true
                Test:           TestRunTable
FAIL
```
with parser_test.go:32 being the `assert` line of the test function. There is no indication of the test that is actually failing.

2. The parser keeps a state. Do you want table-like tests for those too? I'm thinking of tests like [this](https://github.com/canonical/sqlair/pull/3/commits/7e6558f0e038167f8f7a4608a955f9d4296ec5bc#diff-acf51ab254114fd61f15d384a41705347f24cf03efacf3447fd9c42f32b35633R37). Wouldn't this over complicate the function using the test table?

niemeyer replied:
For (1), just print the test line being executed with %#v on the failure cases, so people can easily spot the failure.

For (2), fair enough, let's keep it as it is for now so we can move on. It's repetitive but it's not crazy yet, so let's wait for until/if it bothers us more.

fernape replied:
As of now, all the tests are in a table-like fashion.

niemeyer commented on internal/parse/parser_test.go:
```diff
@@ -12,3 +12,73 @@ func TestInit(t *testing.T) {
+func TestPeekByte(t *testing.T) {
+	p := NewParser()
+	p.Parse("")
+	assert.Equal(t, false, p.peekByte('a'))
```
Might be worth using go-check, which allows us to have suites and is what we use the most at Canonical. I can provide some simple samples if needed.  Okay to wait a bit until it's more obvious, if you feel like it.

niemeyer reviewed (changes requested):
Looks good. Just a comment on the tests.

There are some potential issues in the base that was already merged, but we'll need to review it elsewhere. For example: QueryPart doesn't look like something that should be public.

fernape committed:
Use table defined tests.

Declare a table of functions, inputs and expected outputs.
Used for isNameByte and peekByte at the moment.

fernape committed:
All tests are table-like.

Adapt some of them to the equivalent ones.
Add stringf0 to point to string functions without parameters.

fernape committed:
Remove isNameByte.

The concept is dubious and unclear.
If needed it will be added when the time comes.

fernape committed:
Make failed tests output clearer.

Print the actual sentence that fails along with actual and expected
values.

niemeyer commented on internal/parse/parser_test.go:
```diff
@@ -1,11 +1,68 @@
+
+	{stringf: p.skipString, result: []bool{false}, input: "", data: []string{"a"}},
+	{stringf: p.skipString, result: []bool{true, true}, input: "helloworld", data: []string{"hElLo", "w"}},
+	{stringf: p.skipString, result: []bool{true, true}, input: "hello world", data: []string{"hello", " "}},
```
Nice, thanks! :+1:
